src/sqlfluff/utils/reflow/classes.py

```diff
@@ -0,0 +1,283 @@
+                if seg.pos_marker.templated_file:
+                    next_raw_slice = (
+                        seg.pos_marker.templated_file.raw_slices_spanning_source_slice(
+                            slice(
+                                trailing_whitespace.pos_marker.source_slice.stop,
+                                trailing_whitespace.pos_marker.source_slice.stop,
+                            )
+                        )
+                    )
+                    # If the next slice is literal, that means it's regular code, so
+                    # it's safe to delete the trailing whitespace. If it's anything
+                    # else, it's template code, so don't delete the whitespace because
+                    # it's not REALLY trailing whitespace in terms of the raw source
+                    # code.
+                    if next_raw_slice and next_raw_slice[0].slice_type != "literal":
+                        reflow_logger.debug(
+                            "    !! TW: Skipping apparent trailing whitespace "
+                            "followed by template"
+                        )
+                        continue
+
+                fixes.append(LintFix("delete", trailing_whitespace))
+                reflow_logger.debug("    !! TW: TRAILING WHITESPACE")
+        return fixes
+
+
+@dataclass
+class ReflowSequence:
+    """Class for keeping track of elements in a reflow operation.
+
+    It is assumed that there will be alternating blocks and points
+    (even if some points have no segments). This is validated on
+    construction.
+
+    We assume points on each end because this is the case with a file.
+    """
+
+    elements: Sequence[Union[ReflowBlock, ReflowPoint]]
+    root_segment: BaseSegment
+
+    def __post_init__(self):
+        """Validate integrity"""
+        assert self.elements
+        # Check root segment is shared
+        assert all(elem.root_segment is self.root_segment for elem in self.elements)
+        # Check first and last
+        try:
+            if isinstance(self.elements[0], ReflowPoint):
+                # Check odds are all points
+                assert all(isinstance(elem, ReflowPoint) for elem in self.elements[::2])
+                # Check evens are all blocks
+                assert all(
+                    isinstance(elem, ReflowBlock) for elem in self.elements[1::2]
+                )
+            else:
+                # Check odds are all points
+                assert all(isinstance(elem, ReflowBlock) for elem in self.elements[::2])
+                # Check evens are all blocks
+                assert all(
+                    isinstance(elem, ReflowPoint) for elem in self.elements[1::2]
+                )
+        except AssertionError as err:
+            reflow_logger.exception(
+                "Assertion check on ReflowSequence failed. " "Elements: %s",
+                self.elements,
+            )
+            raise err
```

Comment:
I'd be inclined to do this in a util function 
`def _validate_reflow_sequence(elements: Sequence[Union[ReflowBlock, ReflowPoint]], root_segment: BaseSegment)`

Reply:
why? I'm not particularly against it, I'm just curious.

Reply:
Actually - on reflection now I'm with you. It will be useful later.